cli/src/org/partiql/shell/ShellHighlighter.kt

```diff
@@ -11,32 +11,134 @@
+        val lineIndexesMap = mutableMapOf<Int, Int>()
+        var currentLine = 0
+        lineIndexesMap[currentLine] = -1
+        input.forEachIndexed { index, char ->
+            if (char == '\n') lineIndexesMap[++currentLine] = index
+        }
 
-    private val syntax = SyntaxHighlighter.build(syntaxFile, "PartiQL")
+        // Get Tokens
+        val tokens: List<Token>
+        try {
+            tokens = lexer.tokenize(input)
+        } catch (e: Exception) {
+            return AttributedString(input, AttributedStyle().foreground(AttributedStyle.RED))
+        }
 
-    override fun highlight(reader: LineReader, line: String): AttributedString = syntax.highlight(line)
+        // Replace Token Colors
+        tokens.forEach { builder = createAttributeStringBuilder(it, lineIndexesMap, builder, false) }
+
+        // Parse
+        try {
+            parser.parseAstStatement(input)
+        } catch (e: ParserException) {
+            val column =
+                e.errorContext[Property.COLUMN_NUMBER]?.longValue()?.toInt() ?: return builder.toAttributedString()
+            val lineNumber =
+                e.errorContext[Property.LINE_NUMBER]?.longValue()?.toInt() ?: return builder.toAttributedString()
+            val token = tokens.find {
+                it.span.column.toInt() == column && it.span.line.toInt() == lineNumber
+            } ?: return builder.toAttributedString()
+            builder = createAttributeStringBuilder(token, lineIndexesMap, builder, true)
+        }
+        return builder.toAttributedString()
+    }
 
     override fun setErrorPattern(errorPattern: Pattern?) {}
 
     override fun setErrorIndex(errorIndex: Int) {}
+
+    /**
+     * Based on the [token] type and location, this function will return a replica of the [stringBuilder] with the
+     * token's location having an updated color-scheme (style). If [isFailure] is passed, the style will always be RED.
+     */
+    private fun createAttributeStringBuilder(
+        token: Token,
+        lineIndexes: Map<Int, Int>,
+        stringBuilder: AttributedStringBuilder,
+        isFailure: Boolean
+    ): AttributedStringBuilder {
+        val a = AttributedStringBuilder()
+        val style = when (isFailure) {
+            true -> AttributedStyle().foreground(AttributedStyle.RED)
+            false -> getStyle(token)
+        }
+        val column = token.span.column.toInt()
+        val lineNumber = token.span.line.toInt() - 1
+        val length = token.span.length.toInt()
+        val index = lineIndexes[lineNumber]?.plus(column) ?: return stringBuilder
+        a.append(stringBuilder.subSequence(0, index))
+        a.append(stringBuilder.subSequence(index, index + length), style)
+        a.append(stringBuilder.subSequence(index + length, stringBuilder.length))
+        return a
+    }
+
+    /**
+     * Sets the color and thickness of the string based on the [token] type
+     */
+    private fun getStyle(token: Token): AttributedStyle {
```

Comment:
May be useful to abstract the Token styling to a separate configuration a user can modify and provide through the `ShellConfiguration`. For accessibility (e.g. color-blindness), it may be helpful for a user to modify the Token styling to something other than the defaults. Feel free to create an issue to track since that configurability may be out of scope for this PR.

Reply:
I would agree this is out of scope, but should definitely be a separate issue because this is a great idea

Reply:
ACK. I'll create an issue for highlighting based on user input. Then, we can store the default input in our resources. I'm not going to move the current syntax rules away from the when() since I'd want to clarify exactly what the input should be like, but that can come later. I'll make that a desired outcome of the issue. Will link.

But yeah, I totally agree. That's why I also created the `--monochrome` flag to turn off syntax highlighting if people have issues seeing it.

Reply:
Not sure if this issue has been created yet. Did you want me to create it?